### PR TITLE
Feature/37 allow edit of round forms

### PR DESCRIPTION
**Pull request by todd-dembrey**
This prevents users from changing the assigned form to a round but allows them to continue to edit it.

### CODE REVIEW

File: opentech/apply/funds/admin.py [hunk 1]
```diff
@@ -74,9 +34,34 @@ class LabAdmin(ModelAdmin):
     menu_label = 'Labs'
 
 
+class NoDeletePermission(PermissionHelper):
```
Comment (zerolab) [inline opentech/apply/funds/admin.py]: Will this be useful for other apps/models?

File: opentech/apply/funds/admin.py [hunk 2]
```diff
@@ -74,9 +34,34 @@ class LabAdmin(ModelAdmin):
...
+    def user_can_delete_obj(self, user, obj):
+        return False
+
+
 class ApplicationFormAdmin(ModelAdmin):
     model = ApplicationForm
     menu_icon = 'form'
+    list_display = ('name', 'funds', 'rounds', 'labs')
```
Comment (zerolab) [inline opentech/apply/funds/admin.py]: As discussed, merge funds/rounds/labs into one column "Used by" and indicate with a label what type it is.

File: opentech/apply/funds/edit_handlers.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+class BaseReadOnlyPanel(EditHandler):
+    def render(self):
+        value = getattr(self.instance, self.attr)
+        if callable(value):
+            value = value()
+        return format_html('<div style="padding-top: 1.2em;">{}</div>', value)
+
+    def render_as_object(self):
+        return format_html(
```
Comment (zerolab) [inline opentech/apply/funds/edit_handlers.py]: Should these not be regular templates?

File: opentech/apply/funds/edit_handlers.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    url_name = f'{obj._meta.app_label}_{obj._meta.model_name}_modeladmin_edit'
+    return reverse(url_name, args=(obj.id,))
+
+
+class BaseReadOnlyInlinePanel(BaseReadOnlyPanel):
+    def render(self):
+        values = getattr(self.instance, self.attr).all()
+        return mark_safe(
+            ''.join(
+                '<div style="padding-top: 1.2em;">'
```
Comment (zerolab) [inline opentech/apply/funds/edit_handlers.py]: I'd rather we use classes and inject admin css, than hardcode all of this.

File: opentech/apply/funds/models.py [hunk 5]
```diff
@@ -219,9 +244,29 @@ class Round(SubmittableStreamForm):
                 FieldPanel('start_date'),
                 FieldPanel('end_date'),
             ]),
-        ], heading="Dates")
+        ], heading="Dates"),
+        ReadOnlyPanel('workflow'),
+        ReadOnlyInlinePanel('forms'),
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: Unfortunately this does not indicate what form is used when creating the round.

File: opentech/apply/funds/edit_handlers.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
+from django.forms.utils import pretty_name
+from django.urls import reverse
+from django.utils.html import format_html, mark_safe
+from django.utils.translation import ugettext_lazy as _
+
+from wagtail.wagtailcore.models import Page
+from wagtail.wagtailadmin.edit_handlers import BaseFieldPanel, EditHandler, FieldPanel
+
+
+class BaseReadOnlyPanel(EditHandler):
```
Comment (zerolab) [inline opentech/apply/funds/edit_handlers.py]: I have a feeling the read-only panels may come handy further down the line. My feeling is they should go to separate app

**Comment by zerolab** [comment]
Oh, and the tests are failing with `list index out of range` in test_errors_if_blank_user_data_even_if_logged_in

File: opentech/apply/funds/admin.py (hunk 1 above)
Comment (todd-dembrey) [reply opentech/apply/funds/admin.py]: Perhaps. But only here for now.

File: opentech/apply/funds/models.py (hunk 5 above)
Comment (todd-dembrey) [reply opentech/apply/funds/models.py]: Yeah, It is a tricky one. Perhaps some help text to describe what is happening

File: opentech/apply/funds/edit_handlers.py (hunk 6 above)
Comment (todd-dembrey) [reply opentech/apply/funds/edit_handlers.py]: They will if needed again

File: opentech/apply/funds/models.py (hunk 5 above)
Comment (todd-dembrey) [reply opentech/apply/funds/models.py]: Text added to improve clarity around form behaviour

**Pull request merged**